File: engines/rdf4j/src/main/kotlin/Engine.kt
package es.weso.rudof

import org.eclipse.rdf4j.common.transaction.IsolationLevels
import org.eclipse.rdf4j.model.impl.LinkedHashModel
import org.eclipse.rdf4j.model.vocabulary.RDF4J
import org.eclipse.rdf4j.repository.sail.SailRepository
import org.eclipse.rdf4j.repository.sail.SailRepositoryConnection
import org.eclipse.rdf4j.rio.RDFFormat
import org.eclipse.rdf4j.rio.Rio
import org.eclipse.rdf4j.sail.memory.MemoryStore
import org.eclipse.rdf4j.sail.shacl.ShaclSail
import org.eclipse.rdf4j.sail.shacl.ShaclSailConnection
import org.eclipse.rdf4j.sail.shacl.results.ValidationReport
import java.io.File
import java.io.StringWriter

class Engine : ValidationEngine<ValidationReport> {
    override val name = "rdf4j"

    private lateinit var conn: SailRepositoryConnection
    private var sizeLogged = false

    override fun loadData(dataPath: String, dataFormat: String, shapesPath: String, shapesFormat: String) {
        val shaclSail = ShaclSail(MemoryStore()).apply {
            validationResultsLimitTotal = -1
            validationResultsLimitPerConstraint = -1
        }

        conn = SailRepository(shaclSail)
            .apply { init() }
            .connection

        conn.apply {
            begin(IsolationLevels.NONE)
            add(File(shapesPath), shapesFormat.asFormat(), RDF4J.SHACL_SHAPE_GRAPH)
            commit()

            begin(IsolationLevels.NONE, ShaclSail.TransactionSettings.ValidationApproach.Disabled)
            add(File(dataPath), dataFormat.asFormat())
            commit()
        }

        if (!sizeLogged) {
            println("[$name] Data graph size:   TODO")
            println("[$name] Shapes graph size: TODO")
            sizeLogged = true
        }
    }

    override fun validate(): ValidationReport {
        return conn.let {
            it.begin(IsolationLevels.NONE)
            val report = (it.sailConnection as ShaclSailConnection).revalidate()
            it.commit()
            report
        }
    }

    override fun generateReport(result: ValidationReport): String =
        StringWriter().use { os ->
            val model = LinkedHashModel()
            result.asModel(model)
            Rio.write(model, os, RDFFormat.TURTLE)
            os.toString()
        }

    fun closeConnection() {
        conn.close()
    }
}

File: engines/maplib/engine.py
from maplib import Model, ValidationReport

from validation_engine import ValidationEngine

class Engine(ValidationEngine[ValidationReport]):
    name = "maplib"

    _DATA_GRAPH_IRI = "urn:bench:data"
    _SHAPES_GRAPH_IRI = "urn:bench:shapes"
    # _REPORT_GRAPH_IRI = "urn:bench:validation-report"
    _BASE_IRI = "http://ex.net/"

    def __init__(self) -> None:
        self._model: Model | None = None
        self._size_logged = False

    def load_data(self, data_path: str, data_format: str, shapes_path: str, shapes_format: str) -> None:
        self._model = Model()
        assert self._model is not None

        self._model.read(
            data_path,
            parallel=False, # Makes trouble while parsing prefixes
            graph=self._DATA_GRAPH_IRI,
            base_iri=self._BASE_IRI
        )
        self._model.read(shapes_path,
            parallel=False, # Makes trouble while parsing prefixes
            graph=self._SHAPES_GRAPH_IRI,
            base_iri=self._BASE_IRI
        )

        if not self._size_logged:
            print(f"[{self.name}] Data graph size:   {self._model.size(self._DATA_GRAPH_IRI)}")
            print(f"[{self.name}] Shapes graph size: {self._model.size(self._SHAPES_GRAPH_IRI)}")
            self._size_logged = True

    def validate(self) -> ValidationReport:
        assert self._model is not None
        return self._model.validate(
            data_graph=self._DATA_GRAPH_IRI,
            shape_graph=self._SHAPES_GRAPH_IRI,
            # report_graph=self._REPORT_GRAPH_IRI,
        )

    def generate_report(self, result: ValidationReport) -> str:
        assert self._model is not None
        return "TODO"
        # return self._model.writes(format="turtle", graph=self._REPORT_GRAPH_IRI)
